src/briefcase/integrations/linuxdeploy.py

```diff
@@ -124,6 +132,228 @@ def patch_elf_header(self):
+
+    @property
+    def plugin_id(self):
+        return self.file_name.split(".")[0].split("-")[2]
+
+    @property
+    def file_path(self):
+        return self.command.tools_path / "linuxdeploy_plugins" / self.plugin_id
+
+
+class LinuxDeployGtkPlugin(LinuxDeployPluginBase):
+    plugin_id = "gtk"
+    full_name = "linuxdeploy GTK plugin"
+
+    @property
+    def file_name(self):
+        return "linuxdeploy-plugin-gtk.sh"
+
+    @property
+    def download_url(self):
+        return (
+            "https://raw.githubusercontent.com/linuxdeploy/linuxdeploy-plugin-gtk/"
+            f"master/{self.file_name}"
+        )
+
+
+class LinuxDeployQtPlugin(LinuxDeployPluginBase):
+    plugin_id = "qt"
+    full_name = "linuxdeploy Qt plugin"
+
+    @property
+    def file_name(self):
+        return f"linuxdeploy-plugin-qt-{self.command.host_arch}.AppImage"
+
+    @property
+    def download_url(self):
+        return (
+            "https://github.com/linuxdeploy/linuxdeploy-plugin-qt/"
+            f"releases/download/continuous/{self.file_name}"
+        )
+
+
+class LinuxDeployLocalFilePlugin(LinuxDeployPluginBase):
+    full_name = "user-provided linuxdeploy plugin from local file"
+    install_msg = "Copying user-provided plugin into project"
+
+    def __init__(self, command, plugin_path, bundle_path):
+        self._file_name = plugin_path.name
+        self.local_path = plugin_path.parent
+        self._file_path = bundle_path
+
+        # Call the super last to ensure validation of the filename
+        super().__init__(command)
+
+    @property
+    def file_name(self):
+        return self._file_name
+
+    @property
+    def file_path(self):
+        return self._file_path
+
+    @property
+    def download_url(self):
+        raise RuntimeError("Shouldn't be trying to download a local file plugin")
+
+    def install(self):
+        # Install the plugin by copying from the local path to the bundle
+        # folder. This is required to ensure that the file is available inside
+        # the Docker context.
+        try:
+            self.command.shutil.copy(
+                self.local_path / self.file_name,
+                self.file_path / self.file_name,
+            )
+        except FileNotFoundError:
+            raise BriefcaseCommandError(
+                f"Could not locate linuxdeploy plugin {self.local_path / self.file_name}. "
+                "Is the path correct?"
+            )
+
+
+class LinuxDeployURLPlugin(LinuxDeployPluginBase):
+    full_name = "user-provided linuxdeploy plugin from URL"
+
+    def __init__(self, command, url):
+        self._download_url = url
+
+        url_parts = urlparse(url)
+        self._file_name = url_parts.path.split("/")[-1]
+
+        # Build a hash of the download URL; this hash is used to
+        # idenfity plugins downloaded from different sources
+        self.hash = hashlib.sha256()
+        for part in url_parts:
+            self.hash.update(part.encode("utf-8"))
```

Comment:
Having 2 different URLs produce different hashes was deliberate. The following 3 URLs could all produce different versions of a plugin:

* `http://example.com/release/linuxdeploy-plugin-foobar.sh`
* `http://example.com/dev/linuxdeploy-plugin-foobar.sh`
* `http://example.com/archive/linuxdeploy-plugin-foobar.sh?version=1`

Hashing on just the URL would make the three appear to be the same plugin.

Admittedly, this is an edge case that is unlikely to be an issue in practice, but if it ever comes up, it would be very difficult to diagnose, so I figured we should be proactive. Documenting the why is definitely a good idea, though.

Reply:
Oh wait - I see what you're saying now - just hash the URL literal, rather than using HMAC. Yeah - that's a good idea - we're not depending on any cryptographic properties of the HMAC. I was overthinking things because I had the parsed URL parts handy.